src/dramatiq_abort/middleware.py

```diff
@@ -231,24 +253,54 @@ class _CtypesAbortManager:
 
     def __init__(self, logger: Optional[Logger] = None):
         self.logger = logger or get_logger(__name__, type(self))
-        self.abortables: Dict[str, int] = {}
+        # This lock avoid race between the monitor and a task cleaning up.
+        self.lock = threading.Lock()
+        self.abortable_messages: Dict[str, int] = {}
+        self.abort_requests: Dict[int, Tuple[str, float]] = {}
 
     def add_abortable(self, message_id: str) -> None:
-        self.abortables[message_id] = threading.get_ident()
+        self.abortable_messages[message_id] = threading.get_ident()
 
     def remove_abortable(self, message_id: str) -> None:
-        self.abortables.pop(message_id, None)
+        with self.lock:
+            thread_id = threading.get_ident()
+            saved_thread_id = self.abortable_messages.pop(message_id, None)
+            saved_message_id, _ = self.abort_requests.pop(thread_id, (None, None))
+            assert not saved_thread_id or thread_id == saved_thread_id
+            assert not saved_message_id or saved_message_id == message_id
 
-    def abort(self, message_id: str) -> None:
-        thread_id = self.abortables.pop(message_id, None)
-        # In case the task was done in between the polling and now.
-        if thread_id is None:
-            return  # pragma: no cover
+    def add_abort_request(self, message_id: str, abort_timeout: int = 0) -> None:
+        with self.lock:
+            thread_id = self.abortable_messages.get(message_id, None)
+            if thread_id is None:
+                # If the task finished before we signaled it to abort
+                return
+            self.abort_requests[thread_id] = (
+                message_id,
+                time.monotonic() + abort_timeout / 1000,
+            )
 
-        self.logger.info(
-            "Aborting task. Raising exception in worker thread %r.", thread_id
-        )
-        raise_thread_exception(thread_id, Abort)
+    def abort_pending(self) -> None:
+        with self.lock:
+            toabort = [
+                thread_id
+                for thread_id, (_, abort_time) in self.abort_requests.items()
+                if time.monotonic() >= abort_time
+            ]
+
+            for thread_id in toabort:
+                message_id, abort_time = self.abort_requests.pop(thread_id, ("", None))
+                saved_thread_id = self.abortable_messages.pop(message_id, None)
+                assert saved_thread_id == thread_id
+                if thread_id is None or abort_time is None:
```

Comment:
I don't see how this condition could ever be `True`. We are inside the lock, so `abort_request[thread_id]` is guaranteed to have value.

Reply:
Actually I'm pretty sure that we really only need line 300-303

Reply:
That is true, the `if` really should be removed.

Although we technically only need the last few lines I think it is better to keep the pops. If for whatever reason remove_abortable doesn't run we would end up aborting the wrong task, so remove them from the list just in case.